Sources/IBMSwiftSDKCore/Authentication/IAMAuthenticator.swift

```diff
@@ -70,28 +76,34 @@ class IAMTokenSource: TokenSource {
     }
 
     func getToken(completionHandler: @escaping (String?, RestError?) -> Void) {
-        // request a new access token if the current token is expired
-        guard let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow > 0 else {
-            requestToken {
-                token, error in
-
-                guard let token = token, error == nil else {
-                    completionHandler(nil, error)
-                    return
+        // If we have a token, pass it to the completion handler.
+        if let token = self.token {
+            completionHandler(token, nil)
+            if let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow < 0 {
+                refreshQueue.async{ self.refreshToken() }
+            }
+        } else {
+            fetchQueue.async {
+                // Check if token was obtained by an earlier fetch
+                if let token = self.token {
+                    completionHandler(token, nil)
+                } else {
+                    self.requestToken(completionHandler: completionHandler)
                 }
-                self.token = token.accessToken
-                // Compute refreshDate for token
-                let expirationDate = Date(timeIntervalSince1970: Double(token.expiration))
-                // We want a little buffer to make sure we refresh proactively
-                let buffer = expirationDate.timeIntervalSinceNow * -0.2
-                self.refreshDate = expirationDate.addingTimeInterval(buffer)
-                completionHandler(self.token, nil)
             }
+        }
+    }
+
+    // request a new access token if the refresh time for the current token is passed
+    func refreshToken() {
+        // If refreshDate is set and still in the future, no refresh needed -- just return
+        if let refreshDate = refreshDate, refreshDate.timeIntervalSinceNow > 0 {
             return
         }
 
-        // use the existing, valid access token
-        completionHandler(self.token, nil)
+        // This is running in a serial dispatch queue, so update to refreshDate is serialized
+        self.refreshDate = Date(timeIntervalSinceNow: 60) // Update refreshDate to 60 seconds from now
+        requestToken { (_, _) in /* dummy completion handler */ }
```

Comment:
No completion handler is needed here because `requestToken` now updates `self.token` and `self.refreshDate` with the response from the token service.